### PR TITLE
#15630 validate no dash in field variable and also deny modify field …

### PR DESCRIPTION
**Pull request by erickgonzalez**
…variable via REST

### CODE REVIEW

File: dotCMS/src/integration-test/java/com/dotcms/rest/api/v2/contenttype/FieldResourceTest.java [hunk 1]
```diff
@@ -82,6 +82,75 @@ public static void cleanUpData() throws DotDataException, DotSecurityException {
...
 
+    @Test
+    public void testCreateRelationshipFieldWithDash_Return400() throws Exception {
+        final FieldResource resource = new FieldResource();
+
+        final ContentType contentType = getContentType();
+
+        final String jsonField = "{"+
+                // IDENTITY VALUES
+                "	\"clazz\" : \"com.dotcms.contenttype.model.field.ImmutableRelationshipField\","+
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/rest/api/v2/contenttype/FieldResourceTest.java]: this hardcoded value can be replaced by `ImmutableRelationshipField.class.getCanonicalName()`

File: dotCMS/src/integration-test/java/com/dotcms/rest/api/v2/contenttype/FieldResourceTest.java [hunk 2]
```diff
@@ -82,6 +82,75 @@ public static void cleanUpData() throws DotDataException, DotSecurityException {
...
+
+    @Test
+    public void testUpdateFieldVariable_Return400() throws Exception {
+        final FieldResource resource = new FieldResource();
+
+        final ContentType contentType = getContentType();
+
+        final String jsonField = "{"+
+                // IDENTITY VALUES
+                "	\"clazz\" : \"com.dotcms.contenttype.model.field.ImmutableRelationshipField\","+
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/rest/api/v2/contenttype/FieldResourceTest.java]: this hardcoded value can be replaced by ImmutableRelationshipField.class.getCanonicalName()

File: dotCMS/src/integration-test/java/com/dotcms/rest/api/v2/contenttype/FieldResourceTest.java [hunk 3]
```diff
@@ -82,6 +82,75 @@ public static void cleanUpData() throws DotDataException, DotSecurityException {
...
+                contentType.id(), jsonField.replace("CONTENT_TYPE_ID", contentType.id()), getHttpRequest());
+
+        assertNotNull(response);
+        assertEquals(200, response.getStatus());
+
+        final Map<String, Object> fieldMap = (Map<String, Object>) ((ResponseEntityView) response.getEntity()).getEntity();
+
+        final String jsonFieldUpdate = "{"+
+                // IDENTITY VALUES
+                "	\"clazz\" : \"com.dotcms.contenttype.model.field.ImmutableRelationshipField\","+
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/rest/api/v2/contenttype/FieldResourceTest.java]: same here

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java [hunk 4]
```diff
@@ -157,13 +154,10 @@ public Response getContentTypeFieldById(@PathParam("typeId") final String typeId
             @PathParam("fieldId") final String fieldId, @Context final HttpServletRequest req)
             throws DotDataException, DotSecurityException {
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java]: why was this call `webResource.init` deleted?

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java [hunk 5]
```diff
@@ -184,13 +178,10 @@ public Response getContentTypeFieldByVar(@PathParam("typeId") final String typeI
             @PathParam("fieldVar") final String fieldVar, @Context final HttpServletRequest req)
             throws DotDataException, DotSecurityException {
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java]: same here

File: dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java [hunk 6]
```diff
@@ -132,6 +132,12 @@ public Field save(final Field field, final User user) throws DotDataException, D
 	    		//Do nothing as Starter comes with id but field is unexisting yet
 	    	}
 	    }else {
+	        //This validation should only be for new fields, since the field velocity var name(variable) can not be modified
+            if(UtilMethods.isSet(field.variable()) && !field.variable().matches("^[a-zA-Z0-9]+")) {
+                String errorMessage = "Field velocity var name "+ field.variable() +" contains characters not allowed, here is a suggestion of the variable: " + com.dotmarketing.util.StringUtils.camelCaseLower(field.variable());
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java]: `errorMessage` can be declared final

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java [hunk 7]
```diff
@@ -157,13 +154,10 @@ public Response getContentTypeFieldById(@PathParam("typeId") final String typeId
...
 
-        final InitDataObject initData = this.webResource.init(null, false, req, false, null);
-        final FieldAPI fapi = APILocator.getContentTypeFieldAPI();
-
         Response response = null;
         try {
 
-            Field field = fapi.find(fieldId);
+            Field field = fieldAPI.find(fieldId);
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java]: I think `field` can be declared final

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java [hunk 8]
```diff
@@ -367,14 +360,13 @@ public Response deleteContentTypeFieldByVar(@PathParam("typeId") final String ty
 
         final InitDataObject initData = this.webResource.init(null, false, req, false, null);
         final User user = initData.getUser();
-        final FieldAPI fapi = APILocator.getContentTypeFieldAPI();
 
         Response response = null;
         try {
 
-            Field field = fapi.byContentTypeIdAndVar(typeId, fieldVar);
+            Field field = fieldAPI.byContentTypeIdAndVar(typeId, fieldVar);
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java]: I think `field` can be declared final

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java [hunk 9]
```diff
@@ -669,6 +669,25 @@ public void testValidateRelationshipField_shouldThrowAnException(TestCase testCa
         }
     }
 
+    @Test(expected = DotDataException.class)
```
[the remark this answers is not in the document]
Comment (erickgonzalez) [reply dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java]: done

File: dotCMS/src/integration-test/java/com/dotcms/rest/api/v2/contenttype/FieldResourceTest.java (hunk 1 above)
Comment (erickgonzalez) [reply dotCMS/src/integration-test/java/com/dotcms/rest/api/v2/contenttype/FieldResourceTest.java]: done

File: dotCMS/src/integration-test/java/com/dotcms/rest/api/v2/contenttype/FieldResourceTest.java (hunk 2 above)
Comment (erickgonzalez) [reply dotCMS/src/integration-test/java/com/dotcms/rest/api/v2/contenttype/FieldResourceTest.java]: done

File: dotCMS/src/integration-test/java/com/dotcms/rest/api/v2/contenttype/FieldResourceTest.java (hunk 3 above)
Comment (erickgonzalez) [reply dotCMS/src/integration-test/java/com/dotcms/rest/api/v2/contenttype/FieldResourceTest.java]: done

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java (hunk 4 above)
Comment (erickgonzalez) [reply dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java]: was marked as unused

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java (hunk 7 above)
Comment (erickgonzalez) [reply dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java]: done

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java (hunk 5 above)
Comment (erickgonzalez) [reply dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java]: unused

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java (hunk 8 above)
Comment (erickgonzalez) [reply dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java]: done

File: dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java (hunk 6 above)
Comment (erickgonzalez) [reply dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java]: done

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java (hunk 5 above)
Comment (nollymar) [reply dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java]: but I think this validation was removed https://github.com/dotCMS/core/blob/e152e08cd494d979dddb7a261286c5976a822d8f/dotCMS/src/main/java/com/dotcms/rest/WebResource.java#L453-L454

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java (hunk 4 above)
Comment (erickgonzalez) [reply dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java]: restored

**Review by nollymar — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java [hunk 10]
```diff
@@ -226,15 +220,17 @@ public Response updateContentTypeFieldById(@PathParam("typeId") final String typ
...
 
-                Field currentField = fapi.find(fieldId);
+                final Field currentField = fieldAPI.find(fieldId);
 
                 if (!currentField.id().equals(field.id())) {
 
-                    response = ExceptionMapperUtil.createResponse(null, "Field id '"+ fieldId +"' does not match a field with id '"+ field.id() +"'");
+                    throw new DotDataValidationException("Field id '"+ fieldId +"' does not match a field with id '"+ field.id() +"'");
 
+                } else if(!currentField.variable().equals(field.variable())){
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java]: Let's move this business rule to the api level. It also applies when using the java api directly

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java [hunk 11]
```diff
@@ -270,15 +263,15 @@ public Response updateContentTypeFieldByVar(@PathParam("typeId") final String ty
 
             } else {
 
-                Field currentField = fapi.byContentTypeIdAndVar(typeId, fieldVar);
+                final Field currentField = fieldAPI.byContentTypeIdAndVar(typeId, fieldVar);
 
                 if (!currentField.id().equals(field.id())) {
 
-                    response = ExceptionMapperUtil.createResponse(null, "Field var '"+ fieldVar +"' does not match a field with id '"+ field.id() +"'");
+                    throw new DotDataValidationException("Field var '"+ fieldVar +"' does not match a field with id '"+ field.id() +"'");
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/rest/api/v2/contenttype/FieldResource.java]: let's fix this message and make it look like the other one: "Field id '"+ fieldId +"' does not match a field with id '"+ field.id() +"'"

**Review by dsilvam — APPROVED** [review]

**Pull request merged**